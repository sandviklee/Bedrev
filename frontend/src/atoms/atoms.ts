import { atom } from "recoil";

/**
 * Global state management for all filters/sorts.
 * Done with recoil.
 */

export const placeState = atom({
  key: "place",
  default: sessionStorage.getItem("place") || "",
});

export const searchState = atom({
  key: "search",
  default: sessionStorage.getItem("search") || "",
});

export const minValueState = atom({
  key: "minValue",
  default: sessionStorage.getItem("minValue") || "0",
});

export const maxValueState = atom({
  key: "maxValue",
  default: sessionStorage.getItem("maxValue") || "10000",
});

export const sortState = atom({
  key: "sort",
  default: sessionStorage.getItem("sort") || "asc",
});

export const ratingState = atom({
  key: "rating",
  default: 0,
});
