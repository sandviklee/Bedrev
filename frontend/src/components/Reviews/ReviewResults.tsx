import { gql, useQuery } from "@apollo/client";
import Review from "./Review";
import { ReviewType } from "../../types/types";
import Pagination from "@mui/material/Pagination";
import { useEffect, useState } from "react";
import { ratingState } from "../../atoms/atoms";
import { useRecoilState } from "recoil";

interface ReviewResults {
  bedrift?: string;
}

const ReviewResults = ({ bedrift }: ReviewResults) => {
  const reviewsShown = 4;
  const rating = useRecoilState(ratingState); //Global State management.
  const [currentPage, setCurrentPage] = useState(1);
  const [dimensions, setDimensions] = useState({
    width: window.innerWidth,
    height: window.innerHeight,
  });
  const [reviewAmount, setReviewAmount] = useState(3);

  /**
   * Handles the dimentions of the current page
   */
  const handleResize = () => {
    setDimensions({
      width: window.innerWidth,
      height: window.innerHeight,
    });
  };

  /**
   * Event listener for the page size.
   */
  useEffect(() => {
    window.addEventListener("resize", handleResize, false);
  }, []);

  /**
   * Resets the page to the first page everytime
   * the search result is filtered.
   */
  const resetPage = () => {
    setCurrentPage(1);
  };

  useEffect(() => {
    resetPage();
  }, [rating[0]]);

  /**
   * Changes how many Reviews
   * should be queried at
   * different screen sizes.
   */
  useEffect(() => {
    const largeMonitorWidth = 1400;
    const middleMonitorWidth = 1000;
    const smallMonitorWidth = 750;
    setReviewAmount(
      dimensions.width > largeMonitorWidth
        ? reviewsShown
        : dimensions.width > middleMonitorWidth
        ? reviewsShown - 1
        : dimensions.width > smallMonitorWidth
        ? reviewsShown - 2
        : 1,
    );
    resetPage();
  }, [dimensions.width]);

  /**
   * Query for reviews given its filters and sort.
   */
  const REVIEW_QUERY = gql`
        {
            reviewSok(organisasjonsnummer: ${parseInt(
              bedrift || "0",
            )}, amount: ${reviewAmount} , skip: ${
              currentPage > 0 ? (currentPage - 1) * reviewAmount : 0
            } ${rating[0] ? `, rating: ${rating[0]}` : ""}) {
                title
                description
                rating
                id
            }
            reviewSokCount(organisasjonsnummer:  ${parseInt(bedrift || "0")} ${
              rating[0] ? `, rating: ${rating[0]}` : ""
            })
        }
    `;

  /**
   * Fetching of said query.
   */
  const { data, loading, error, refetch } = useQuery(REVIEW_QUERY);

  if (loading) return <p>Loading...</p>;
  if (error) return <pre>{error.message}</pre>;

  /**
   * Function to update the current page.
   * @param event for handling page change events
   * @param value for getting the value from pagination
   */
  const handleChange = (event: React.ChangeEvent<unknown>, value: number) => {
    event;
    setCurrentPage(value);
    refetch();
  };

  return (
    <>
      <div className="sm:pt-6 pt-2 flex flex-row gap-4 sm:justify-start justify-center ">
        {data["reviewSok"].length > 0 ? (
          data["reviewSok"].map((review: ReviewType) => <Review {...review} />)
        ) : (
          <p className="sm:text-lg text-base pb-6">
            Denne bedrifter har ingen tilbakemeldinger enda med gitte filter,
            jobber du her?{" "}
            <span className="underline">Skriv en tilbakemelding</span>
          </p>
        )}
      </div>
      <div className="flex justify-center pt-8">
        <Pagination
          page={currentPage}
          onChange={handleChange}
          onClick={refetch}
          count={Math.ceil(data["reviewSokCount"] / reviewAmount)}
        />
      </div>
    </>
  );
};

export default ReviewResults;
