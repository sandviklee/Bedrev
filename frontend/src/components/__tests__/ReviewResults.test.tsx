import { render, screen, waitFor } from "@testing-library/react";
import { MockedProvider } from "@apollo/client/testing";
import { gql } from "@apollo/client";
import ReviewResults from "../Reviews/ReviewResults";
import { RecoilRoot, MutableSnapshot } from "recoil";
import { ratingState } from "../../atoms/atoms";
import { BrowserRouter as Router } from "react-router-dom";

const REVIEW_QUERY = gql`
    {
        reviewSok(organisasjonsnummer: ${0}, amount: ${3}, skip: ${0}, rating: ${5}) {
            title
            description
            rating
            id
        }
        reviewSokCount(organisasjonsnummer: ${0}, rating: ${5})
    }
`;

const mocks = [
  {
    request: {
      query: REVIEW_QUERY,
    },
    result: {
      data: {
        reviewSok: [
          {
            title: "Test Review",
            description: "This is a test review.",
            rating: 5,
            id: 0,
          },
        ],
        reviewSokCount: 1,
      },
    },
  },
];

const initializeState = (mutableSnapshot: MutableSnapshot) => {
  mutableSnapshot.set(ratingState, 5);
};

describe("ReviewResults component", () => {
  test("should render loading state initially", () => {
    const { getByText } = render(
      <Router>
        <RecoilRoot initializeState={initializeState}>
          <MockedProvider mocks={[]}>
            <ReviewResults bedrift="0" />
          </MockedProvider>
        </RecoilRoot>
      </Router>,
    );

    expect(getByText("Loading...")).toBeInTheDocument();
  });

  test("should render reviews", async () => {
    render(
      <Router>
        <RecoilRoot initializeState={initializeState}>
          <MockedProvider mocks={mocks} addTypename={false}>
            <ReviewResults bedrift="0" />
          </MockedProvider>
        </RecoilRoot>
      </Router>,
    );

    await waitFor(() => screen.getByText("Test Review"));

    expect(screen.getByText("Test Review")).toBeInTheDocument();
    expect(screen.getByText("This is a test review.")).toBeInTheDocument();
  });

  test("matches snapshot", async () => {
    const { asFragment } = render(
      <Router>
        <RecoilRoot initializeState={initializeState}>
          <MockedProvider mocks={mocks} addTypename={false}>
            <ReviewResults bedrift="0" />
          </MockedProvider>
        </RecoilRoot>
      </Router>,
    );

    expect(await asFragment()).toMatchSnapshot();
  });
});
