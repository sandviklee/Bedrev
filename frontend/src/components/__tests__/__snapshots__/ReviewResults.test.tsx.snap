// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ReviewResults component > matches snapshot 1`] = `
<DocumentFragment>
  <p>
    Loading...
  </p>
</DocumentFragment>
`;
